modules/imgproc/src/warp_kernels.simd.hpp

```diff
@@ -2130,6 +2195,2067 @@ void warpAffineLinearApproxInvoker_8UC4(const uint8_t *src_data, size_t src_step
+                    p01r = vreinterpret_u8_u32(vtrn2_u32(p00_, p01_));
+                    p10r = vreinterpret_u8_u32(vtrn1_u32(p10_, p11_));
+                    p11r = vreinterpret_u8_u32(vtrn2_u32(p10_, p11_));
+
+                    p00_ = vreinterpret_u32_u8(vtbl4_u8(p00, greens));
+                    p01_ = vreinterpret_u32_u8(vtbl4_u8(p01, greens));
+                    p10_ = vreinterpret_u32_u8(vtbl4_u8(p10, greens));
+                    p11_ = vreinterpret_u32_u8(vtbl4_u8(p11, greens));
+
+                    p00g = vreinterpret_u8_u32(vtrn1_u32(p00_, p01_));
+                    p01g = vreinterpret_u8_u32(vtrn2_u32(p00_, p01_));
+                    p10g = vreinterpret_u8_u32(vtrn1_u32(p10_, p11_));
+                    p11g = vreinterpret_u8_u32(vtrn2_u32(p10_, p11_));
+
+                    p00_ = vreinterpret_u32_u8(vtbl4_u8(p00, blues));
+                    p01_ = vreinterpret_u32_u8(vtbl4_u8(p01, blues));
+                    p10_ = vreinterpret_u32_u8(vtbl4_u8(p10, blues));
+                    p11_ = vreinterpret_u32_u8(vtbl4_u8(p11, blues));
+
+                    p00b = vreinterpret_u8_u32(vtrn1_u32(p00_, p01_));
+                    p01b = vreinterpret_u8_u32(vtrn2_u32(p00_, p01_));
+                    p10b = vreinterpret_u8_u32(vtrn1_u32(p10_, p11_));
+                    p11b = vreinterpret_u8_u32(vtrn2_u32(p10_, p11_));
+
+                    p00_ = vreinterpret_u32_u8(vtbl4_u8(p00, alphas));
+                    p01_ = vreinterpret_u32_u8(vtbl4_u8(p01, alphas));
+                    p10_ = vreinterpret_u32_u8(vtbl4_u8(p10, alphas));
+                    p11_ = vreinterpret_u32_u8(vtbl4_u8(p11, alphas));
+
+                    p00a = vreinterpret_u8_u32(vtrn1_u32(p00_, p01_));
+                    p01a = vreinterpret_u8_u32(vtrn2_u32(p00_, p01_));
+                    p10a = vreinterpret_u8_u32(vtrn1_u32(p10_, p11_));
+                    p11a = vreinterpret_u8_u32(vtrn2_u32(p10_, p11_));
+                } else {
+                    CV_WARP_LINEAR_VECTOR_SHUFFLE_NOTALLWITHIN(C4, 8U);
+
+                    p00r = vld1_u8(pixbuf);
+                    p01r = vld1_u8(pixbuf + 8);
+                    p10r = vld1_u8(pixbuf + 16);
+                    p11r = vld1_u8(pixbuf + 24);
+
+                    p00g = vld1_u8(pixbuf + 32);
+                    p01g = vld1_u8(pixbuf + 32 + 8);
+                    p10g = vld1_u8(pixbuf + 32 + 16);
+                    p11g = vld1_u8(pixbuf + 32 + 24);
+
+                    p00b = vld1_u8(pixbuf + 64);
+                    p01b = vld1_u8(pixbuf + 64 + 8);
+                    p10b = vld1_u8(pixbuf + 64 + 16);
+                    p11b = vld1_u8(pixbuf + 64 + 24);
+
+                    p00a = vld1_u8(pixbuf + 96);
+                    p01a = vld1_u8(pixbuf + 96 + 8);
+                    p10a = vld1_u8(pixbuf + 96 + 16);
+                    p11a = vld1_u8(pixbuf + 96 + 24);
+                }
+
+                v_float16 f00r = v_float16(vcvtq_f16_u16(vmovl_u8(p00r)));
```

Comment:
I guess, this code is a duplicate with warpAffine and there will be another duplicate with remap. Please, create macros or, if possible, inline functions for different types for bilinear interpolation and reuse them. We will have to maintain this code for a long time. Also, I expect that on Intel another implementation of bilinear (bicubic etc.) interpolation may be more efficient. So, let's compact the code. That is, refactored warpAffine should be also made a part of this patch.

Reply:
> refactored warpAffine should be also made a part of this patch.

Done with macros.

> Intel another implementation of bilinear (bicubic etc.)

Which one do you refer to?